src/Active.hs

```diff
@@ -0,0 +1,1670 @@
+--   first.
+--
+--   <<diagrams/src_Active_matchDurationDia.svg#diagram=matchDurationDia&width=450>>
+--
+--   > mdExA :: ActF (Sum Double)
+--   > mdExA = movie [ interval 1 3 ] <#> fromRational <#> Sum
+--   >
+--   > mdExB :: ActF (Sum Double)
+--   > mdExB = (cos'/3) # cut 5 <#> Sum
+--   >
+--   > mdEx :: ActF (Sum Double)
+--   > mdEx = matchDuration mdExA mdExB
+
+matchDuration :: Active 'F a -> Active 'F b -> Active 'F b
+matchDuration (Active (Duration d1) _) b@(Active (Duration d2) _) = stretch (d1/d2) b
+
+-- > matchDurationDia = drawChain
+-- >   [ illustrateActives' 0.1 [[],[]] $ map (<#> getSum) [mdExA, mdExB]
+-- >   , illustrateActive' 0.1 [] $ mdEx <#> getSum
+-- >   ]
+
+
+-- | Stretch a finite active by whatever factor is required so that it
+--   ends up with the given duration.
+--
+--   <<diagrams/src_Active_stretchToDia.svg#diagram=stretchToDia&width=200>>
+--
+--   > stretchToEx :: ActF Rational
+--   > stretchToEx = interval 0 3 # stretchTo 5
+--
+--   >>> durationF (stretchTo 5 (interval 0 3))
+--   5 % 1
+stretchTo :: Rational -> Active 'F a -> Active 'F a
+stretchTo n a@(Active (Duration d) _) = stretch (n / d) a
+
+-- > stretchToDia = illustrateActive stretchToEx
+
+
+-- | Take a "snapshot" of a given 'Active' at a particular time,
+--   freezing the resulting value into an infinite constant.
+--
+--   @snapshot t a = always (runActive a t)@
+--
+--   <<diagrams/src_Active_snapshotDia.svg#diagram=snapshotDia&width=450>>
+--
+--   > snapshotEx :: ActI Double
+--   > snapshotEx = snapshot (1/8) cos'
+--
+--   >>> take 3 (samples 1 (snapshot (9/8) cos'))
+--   [0.7071067811865477,0.7071067811865477,0.7071067811865477]
+snapshot :: Rational -> Active f a -> Active 'I a
+snapshot t a = always (runActive a t)
+
+-- > snapshotDia = illustrateActiveFun' 0.1 [(9/8,CC)] [] (snapshot (1/8)) cos'
+
+
+-- | @cut d a@ cuts the given 'Active' @a@ (which can be finite or
+--   infinite) to the specified finite duration @d@.  Has no effect if
+--   @a@ is already shorter than @d@.  This is analogous to 'take' on
+--   lists.
+--
+--   @cut d dur = interval 0 d@
+--
+--   <<diagrams/src_Active_cutDia.svg#diagram=cutDia&width=450>>
+--
+--   > cutEx :: ActF Double
+--   > cutEx = cut 1.7 cos'
+--
+--   >>> samples 1 (cut 2 dur)
+--   [0 % 1,1 % 1,2 % 1]
+--   >>> samples 1 (cut 3 (interval 0 1))
+--   [0 % 1,1 % 1]
+cut :: Rational -> Active f a -> Active 'F a
+cut c (Active d f) = Active (Duration c `minDuration` d) f
+
+-- > cutDia = illustrateActiveFun' 0.1 [] [] (cut 1.7) cos'
+
+
+-- | @cutTo a1 a2@ cuts @a2@ to match the duration of @a1@, unless
+--   @a2@ is already shorter than @a1@ in which case @cutTo a1 = id@.
+--
+--   @cutTo = cut . durationF@
+cutTo :: Active 'F a -> Active f a -> Active 'F a
+cutTo = cut . durationF
+
+
+-- | @omit d a@ omits the first @d@ time units from @a@. The result is
+--   only defined if @d@ is less than or equal to the duration of @a@.
```

Comment:
I agree, using `end` here doesn't make much sense. You could use `mempty` instead of `error`.